### PR TITLE
Snapshot design

### PR DESCRIPTION
**Pull request by obnoxxx**
### What does this PR achieve? Why do we need it?

This is the proposed design for volume snapshot and clone operations.

### CODE REVIEW

File: docs/design/snapshot.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+# API Proposal
+
+The API is layed out here for file volume types only. The samy API will be
+added for block-volumes at a later time.
+
+### Create a Snapshot
+* **Method:** _POST_
+* **Endpoint**:`/volumes/{volume_uuid}/snapshots`
+* **Content-Type**: `application/json`
+* **Response HTTP Status Code**: 202, See [Asynchronous Operations](#asynchronous-operations)
```
Comment (humblec) [inline docs/design/snapshot.md]: I think this need not be an async call, any specific reason for making it async call?

**Comment by obnoxxx** [comment]
@raghavendra-talur @phlogistonjohn here's the current snapshot proposal.

One question about the cli:
Would you rather have it structured like currently, i.e.
```
heketi-cli volume snapshot-create ID
heketi-cli volume snapshot-delete ID
heketi-cli volume snapshot-list
...
```

or rather have a sub-hierarchy like this

```
heketi-cli volume snapshot [--help]
heketi-cli volume snapshot create ID
heketi-cli volume snapshot delete ID
...
```

Opinions?

This is mostly cosmetic. I personally tend towards the second option.

**Comment by phlogistonjohn** [comment]
@humblec I think every call that uses the heketi executor should be async. It's both consistent with the previous implementation of calls and avoids any http issues in case the operation on the gluster side is slower than we expect. So even if he operation in gluster normally is fast its "safer" to make it async.

**Comment by humblec** [comment]
@phlogistonjohn Oh.. if all the calls are async , then its fine. thanks :+1:

File: docs/design/snapshot.md (hunk 1 above)
Comment (nixpanic) [reply docs/design/snapshot.md]: Gluster needs to do some work on different storage servers, and that may require some time. Doing this async seems appropriate?

**Comment by humblec** [comment]
one last question, did we change volume create call to async recently? or is it async from start?

File: docs/design/snapshot.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+$ heketi-cli volume snapshot-create <vol_uuid> [--name=<snap_uuid>] [--description=<string>]
+$ heketi-cli volume snapshot-clone <snap_uuid> [--name=<vol_uuid>]
+$ heketi-cli volume snapshot-delete <snap_uuid>
+$ heketi-cli volume snapshot-list <vol_uuid>
+$ heketi-cli volume snapshot-info <snap_uuid>
+```
+
+# API Proposal
+
+The API is layed out here for file volume types only. The samy API will be
```
Comment (humblec) [inline docs/design/snapshot.md]: s/samy/same

**Comment by obnoxxx** [comment]
@humblec 
> one last question, did we change volume create call to async recently? or is it async from start?

From the start.

File: docs/design/snapshot.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+```json
+{
+    "name": "midnight",
+    "description": "nightly snapshot"
+}
+```
+
+### Clone a Volume from a Snapshot
+* **Method:** _POST_
+* **Endpoint**:`/volumes/{volume_uuid}/snapshots/{snapshot_uuid}/clone`
```
Comment (humblec) [inline docs/design/snapshot.md]: Do we have a cli for this ?

File: docs/design/snapshot.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    "id": "70927734601288237463aa",
+    "name": "midnight",
+    "description": "nightly snapshot",
+    "created": "1518712323",
+    "volume": "aa927734601288237463aa",
+    "cluster": "67e267ea403dfcdf80731165b300d1ca"
+}
+```
+
+### Clone a Volume directly
```
Comment (humblec) [inline docs/design/snapshot.md]: Do we have a cli for this ?

File: docs/design/snapshot.md (hunk 2 above)
Comment (obnoxxx) [reply docs/design/snapshot.md]: thx. will update

File: docs/design/snapshot.md (hunk 4 above)
Comment (obnoxxx) [reply docs/design/snapshot.md]: yes

File: docs/design/snapshot.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+original. The cloned volumes can be deleted with the `heketi-cli volume delete`
+command. In a similar fashion, snapshots can be removed with the `heketi-cli
+volume snapshot-delete` command.
+
+# Proposed CLI
+
+```
+$ heketi-cli volume snapshot-create <vol_uuid> [--name=<snap_uuid>] [--description=<string>]
+$ heketi-cli volume snapshot-clone <snap_uuid> [--name=<vol_uuid>]
+$ heketi-cli volume snapshot-delete <snap_uuid>
```
Comment (humblec) [inline docs/design/snapshot.md]: Can we remove all snapshots via `--all` or similar switch ?

File: docs/design/snapshot.md (hunk 4 above)
Comment (humblec) [reply docs/design/snapshot.md]: I dont see it in the doc, may be my bad eyes. :)

File: docs/design/snapshot.md (hunk 5 above)
Comment (obnoxxx) [reply docs/design/snapshot.md]: Interesting thought. Not planned yet, but worth a thought. (Maybe for a future addition.)

File: docs/design/snapshot.md (hunk 4 above)
Comment (obnoxxx) [reply docs/design/snapshot.md]: Updated the doc to mention it. (we had it in the started code already)

File: docs/design/snapshot.md (hunk 2 above)
Comment (obnoxxx) [reply docs/design/snapshot.md]: fixed and update

File: docs/design/snapshot.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
+Snapshots are read-only copies of volumes (once the snapshot is activated) that
+can be used for cloning new volumes from. Its is possible to create a snapshot
+through the [Snapshot Create API](../api/api.md#create-a-snapshot) or the
+commandline client.
+
+From the command line client, you can type the following to create a snapshot
+from an existing volume:
+
+```
+$ heketi-cli volume snapshot-create <vol_uuid> [-name=<snap_name>]
```
Comment (phlogistonjohn) [inline docs/design/snapshot.md]: Do you want "--name" for consistency with other commands?

**Comment by phlogistonjohn** [comment]
```
heketi-cli volume snapshot [--help]
heketi-cli volume snapshot create ID
heketi-cli volume snapshot delete ID
...
```

> Opinions?
>
> This is mostly cosmetic. I personally tend towards the second option.

I strongly prefer the 2nd form as well. In fact if we do have some reason not to allow "three tiered" commands I'd put the dash between volume and snapshot rather than snapshot and VERB.
IOW: "heketi-cli volume-snapshot create ID". But again, I'd rather have them all be space separated words like in your 2nd example.

File: docs/design/snapshot.md (hunk 6 above)
Comment (nixpanic) [reply docs/design/snapshot.md]: Yes, that was the idea. However `snap_name` is something that can only be used once in a Gluster environment (where many different volumes may exist). So we'll keep the name in the Heketi DB only, and will not pass it on the `gluster` cmdline.

If there is no value in having a name, we can drop it.

**Comment by nixpanic** [comment]
I've updated the design

1. to match the `heketi-cli volume snapshot <cmd>` format (was `heketi-cli volume snapshot-<cmd>`
1. with a note about a possible future extension to delete all snapshots from a volume (but a `--all` option might be inappropriate)

**Comment by obnoxxx** [comment]
The only concern I have with the current design of the snapshot API and CLI is that it is slightly inconsistent in that a snapshot can be identified by its own UUID, but the operations acting on the snapshot still need to specify the volume id of which the snapshot was taken in the API, e.g. this is the url for operating on the snapshot with uuid snapshot_uuid
```
http(s)://heketi-address:heketiport/volumes/{volume_uuid}/snapshots/{snapshot_uuid}
```
while the corresponding cli would be
```
heketi-cli volume snapshot clone snapshot_uuid
heketi-cli volume snapshot delete snapshot_uuid
```
etc.

**Comment by nixpanic** [comment]
Indeed, the `volume_uuid` is not needed for all snapshot operations, only for `create` and maybe `list`.

We could use `/snapshots/{snapshot_uuid}/...` for the 'pure' snapshot commands. But, when we have snapshots for block-volumes, we'll need to figure out how that fits in the grand scheme of things. I'm happy to change the endpoint to `/snapshots/` and in future we can figure out what type the `{snapshot_uuid}` refers to.

Opinions?

**Comment by nixpanic** [comment]
The [workflow with the latest change](https://github.com/obnoxxx-collab/heketi/blob/dd7f23f6aa522e1acee71594c7430f1fe176cf73/docs/design/snapshot.md) might be cleaner. It boils down to having these two `volume` commands:

1. `heketi-cli volume clone <volume_uuid>`
2. `heketi-cli volume snapshot <volume_uuid>`

The 2nd command creates a snapshot object that can be manipulated through `heketi-cli snapshot ...` commands (with endpoint `/snapshots/{snapshot_uuid}`).

**Comment by obnoxxx** [comment]
@nixpanic I like this a lot, thanks!

**Comment by phlogistonjohn** [comment]
> I'm happy to change the endpoint to /snapshots/ and in future we can figure out what type the {snapshot_uuid} refers to.

Maybe this is a silly question, but does the block volume do anything special with snapshots? If not, then I see no issues with this.

**Comment by obnoxxx** [comment]
@phlogistonjohn wrote:
> >    I'm happy to change the endpoint to /snapshots/ and in future we can figure out what type the {snapshot_uuid} refers to.
> 
> Maybe this is a silly question, but does the block volume do anything special with snapshots? If not, then I see no issues with this.

Not silly at all. On the api side this would just be the same.
In the backend, it would call out for different operations to create/work on the snapshots for volumes and block-volumes. I.e. there would need to be some branching of the form
```
if type == file-volume {
    do_the_file_volume_snapshot_operation()
} else {
    do_the_block_volume_snapshot_operation()
}
```
The reason is that the volume and blockvolume apis are separate.
I think this would be ok.

File: docs/design/snapshot.md [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+
+# API Proposal
+
+The API is layed out here for file volume types only. The same API will be
+added for block-volumes at a later time.
+
+### Create a Snapshot
+* **Method:** _POST_
+* **Endpoint**:`/volumes/{volume_uuid}/snapshot`
```
Comment (nixpanic) [inline docs/design/snapshot.md]: @obnoxxx this lists `/volumes/{volume_uuid}/snapshot`, in your WIP implementation you used `.../snapshots` (plural). Which would be best? Maybe at one point we would like to list all snapshots of a particular volume (that has been omitted in this design for now). The proposed "Remove all Snapshots of a Volume" might suggest that `.../snapshots` is more suitable.

**Comment by nixpanic** [comment]
This PR now has 11 commits that make the initial proposal to a pretty much final form. There is no description of the database contents yet. That will likely get added later once the implementation progresses and we have a better idea what is needed.

Do a :+1: on this comment and I'll squash all the patches into a single one that can be reviewed for merging. Thanks!

File: docs/design/snapshot.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
+Snapshots are read-only copies of volumes (once the snapshot is activated) that
+can be used for cloning new volumes from. Its is possible to create a snapshot
+through the [Snapshot Create API](#create-a-snapshot) or the commandline
+client.
+
+From the command line client, you can type the following to create a snapshot
+from an existing volume:
+
+```
+$ heketi-cli volume snapshot <vol_uuid> [-name=<snap_name>]
```
Comment (phlogistonjohn) [inline docs/design/snapshot.md]: Should probably be a double dash (--name)

File: docs/design/snapshot.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+```
+
+The new snapshot can be used to create a new volume:
+
+```
+$ heketi-cli snapshot clone <snap_uuid> [-name=<clone_name>]
```
Comment (phlogistonjohn) [inline docs/design/snapshot.md]: Another single dash.

File: docs/design/snapshot.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+```json
+{
+    "id": "70927734601288237463aa",
+    "name": "midnight",
+    "description": "nightly snapshot",
+    "created": "1518712323",
+    "volume": "aa927734601288237463aa",
+    "cluster": "67e267ea403dfcdf80731165b300d1ca"
+    "type": "volume"
```
Comment (phlogistonjohn) [inline docs/design/snapshot.md]: The line above lacks a comma

File: docs/design/snapshot.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
+Snapshots are read-only copies of volumes (once the snapshot is activated) that
+can be used for cloning new volumes from. Its is possible to create a snapshot
```
Comment (phlogistonjohn) [inline docs/design/snapshot.md]: The word from at the end of the first sentence is awkward and not needed.

"Its is" should be corrected.

**Comment by nixpanic** [comment]
Thanks John! All remarks should have been resolved now.

@heketi/maintainers please review this (hopefully) final document.

**Comment by nixpanic** [comment]
@raghavendra-talur do you any objections or changes you would like to see?

**Pull request merged**